Libplanet/Net/Protocols/KademliaProtocol.cs

```diff
@@ -708,53 +714,35 @@ private void ReceivePing(Ping ping)
             {
                 await Task.WhenAll(awaitables);
             }
-            catch (AggregateException e)
+            catch (Exception e)
             {
-                if (e.InnerExceptions.All(ie => ie is TimeoutException) &&
-                    e.InnerExceptions.Count == awaitables.Length)
-                {
-                    throw new TimeoutException(
-                        $"All neighbors found do not respond in {_requestTimeout}."
-                    );
-                }
-
+                // FIXME: If timeout occurred, the peer should be removed from the peers list.
                 _logger.Error(
                     e,
                     "Some responses from neighbors found unexpectedly terminated: {Exception}",
                     e
                 );
             }
 
-            for (int i = 0; i < closestNeighbors.Count; i++)
+            var findNeighboursTasks = new List<Task<BoundPeer>>();
+            var count = 0;
+            foreach (var peer in peers)
             {
-                if (string.CompareOrdinal(
-                        closestNeighbors[i].Address.ToHex(),
-                        searchAddress.ToHex()
-                   ) == 0 && _routing.Contains(closestNeighbors[i]))
+                if (history.Contains(peer))
                 {
-                    peerFound = closestNeighbors[i];
-                    return peerFound;
+                    continue;
                 }
-            }
 
-            var findNeighboursTasks = new List<Task<BoundPeer>>();
-            Peer closestKnown = closestNeighbors.Count == 0 ? null : closestNeighbors[0];
-            for (int i = 0; i < Kademlia.FindConcurrency && i < peers.Count; i++)
-            {
-                if (peerFound is null ||
-                    string.CompareOrdinal(
-                        Kademlia.CalculateDistance(peers[i].Address, target).ToHex(),
-                        Kademlia.CalculateDistance(closestKnown.Address, target).ToHex()
-                    ) < 1)
+                findNeighboursTasks.Add(FindSpecificPeerAsync(
+                    history,
+                    target,
+                    peer,
+                    (depth == -1) ? depth : depth - 1,
+                    timeout,
+                    cancellationToken));
+                if (count++ >= Kademlia.FindConcurrency)
                 {
-                    findNeighboursTasks.Add(FindSpecificPeerAsync(
-                        history,
-                        target,
-                        peers[i],
-                        (depth == -1) ? depth : depth - 1,
-                        searchAddress,
-                        timeout,
-                        cancellationToken));
+                    break;
```

Comment:
Would it be difficult to have a unit test for this case?

Reply:
This if statement was not appropriate for the purpose of this method... So I decided to delete this condition.